src/pepr/operator/controllers/monitoring/service-monitor.ts

```diff
@@ -0,0 +1,86 @@
+import { K8s, Log } from "pepr";
+
+import { Prometheus, UDSPackage } from "../../crd";
+import { Monitor } from "../../crd/generated/package-v1alpha1";
+import { getOwnerRef, sanitizeResourceName } from "../utils";
+
+/**
+ * Generate a service monitor for a service
+ *
+ * @param pkg UDS Package
+ * @param namespace
+ */
+export async function serviceMonitor(pkg: UDSPackage, namespace: string) {
+  const pkgName = pkg.metadata!.name!;
+  const generation = (pkg.metadata?.generation ?? 0).toString();
+
+  // Get the list of monitored services
+  const monitorList = pkg.spec?.monitor ?? [];
+
+  // Create a list of generated ServiceMonitors
+  const payloads: Prometheus.ServiceMonitor[] = [];
+
+  for (const monitor of monitorList) {
+    const { selector, portName } = monitor;
+    const name = generateSMName(pkg, monitor);
+    const endpoints: Prometheus.Endpoint[] = [
+      {
+        port: portName,
+        path: monitor.path || "/metrics",
+      },
+    ];
+    const promSelector: Prometheus.Selector = {
+      matchLabels: selector,
+    };
+    const payload: Prometheus.ServiceMonitor = {
+      metadata: {
+        name,
+        namespace,
+        labels: {
+          "uds/package": pkgName,
+          "uds/generation": generation,
+        },
+        ownerReferences: getOwnerRef(pkg),
+      },
+      spec: {
+        endpoints: endpoints,
+        selector: promSelector,
+      },
+    };
+
+    // Apply the VirtualService and force overwrite any existing policy
+    await K8s(Prometheus.ServiceMonitor).Apply(payload, { force: true });
```

Comment:
Not going to hold up this PR but if we could do our best to avoid mixing the code for "resource building" and us using KFC to apply it will make unit testing this code a lot easier.  

It is a small thing but wherever we can segregate pure functions we should try to.

Reply:
Lmk if this fits what you expected - https://github.com/defenseunicorns/uds-core/pull/256/commits/9cbd7db3cc8334930761c52b6c52c631fe79c86d - unit test is a bit eh, but attempted to just pull that generation piece out. Think I could've recreated/grabbed some of the vars in the function itself rather than passing them, but they're needed in both spots so it seemed 50/50.